### ISSUE TITLE
Follow-up on #10482, backwards incomaptible change to type narrowing with TypedDict

### ISSUE THREAD

**Issue by kylebebak** [microsoft/pyright#10493]
**Describe the bug**

Upgrading from 1.1.400 to 1.1.401 causes `pyright` to flag code that relies on [`S in D`](https://github.com/microsoft/pyright/releases/tag/1.1.401) type narrowing in `TypedDict` that it previously did not flag. I'm not sure if this is an issue with `pyright`, or with https://peps.python.org/pep-0728/, or with neither, but it's surprising, and I can't make sense of it after reading through https://github.com/microsoft/pyright/issues/10482 and looking a bit at the PEP

**Code**

[This following code is OK with 1.1.400](https://pyright-play.net/?pyrightVersion=1.1.400&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFD1UA2AhgM6tQCCAFEQmZRoBKAFy0oEqMxFRWMEOMnNps%2BfVpM2HAEK8SFajFGKJAIxlyFkqKfOqF9UsWBRgKbqRmcoAHyjbjayQXACJmEMwUKE8TaygQYhgAVxAo0gBtMPCAXVjgqBDTCNRosTjJBOTU6MzbENygA), but the same code [is not OK with 1.1.401 if experimental features aren't enabled](https://pyright-play.net/?pyrightVersion=1.1.401&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFD1UA2AhgM6tQCCAFEQmZRoBKAFy0oEqMxFRWMEOMnNps%2BfVpM2HAEK8SFajFGKJAIxlyFkqKfOqF9UsWBRgKbqRmcoAHyjbjayQXACJmEMwUKE8TaygQYhgAVxAo0gBtMPCAXVjgqBDTCNRosTjJBOTU6MzbENygA), i.e. this is a backwards incompatible change in pyright's behavior

```python
from typing import TypedDict


class A(TypedDict):
    a: str
    aa: str


class B(TypedDict):
    b: str
    bb: str


def fn(d: A | B):
    if "a" in d:
        return d["aa"]
    if "b" in d:
        return d["bb"]
```

---

[The following code is OK with 1.1.401 and `"enableExperimentalFeatures": true`](https://pyright-play.net/?pyrightVersion=1.1.401&enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFD1UA2AhgM6tQCCAFEQmZRoAaKEzCsyAXgIgArsQCUALlpQ1UZkqisYIVeuabtu%2BrSZsOAIV4kK1GCLETS0uYpXqoAIy069nrx9jPXpSYmAoYBRuUi1OKAAfKEtlfTUkCIAiZkzMFChYtM8oEGIYWRB80gBtbJyAXSKMqEyvXNQCj2K1UvLKgtrAzMagA), but pyright users likely don't want to have to enable experimental features and change the way they define `TypedDict`s to get the same type narrowing behavior as they had in previous `pyright` versions

```python
from typing import TypedDict


class A(TypedDict, closed=True):
    a: str
    aa: str


class B(TypedDict, closed=True):
    b: str
    bb: str


def fn(d: A | B):
    if "a" in d:
        return d["aa"]
    if "b" in d:
        return d["bb"]
```

Shouldn't pyright, and PEP-728, treat `closed=True` as the default for `TypedDict`s? [The PEP says this](https://peps.python.org/pep-0728/#the-closed-class-parameter):

> When neither `extra_items` nor `closed=True` is specified, `closed=False` is assumed

[It also says this about backwards compat](https://peps.python.org/pep-0728/#backwards-compatibility):

> Because extra_items is an opt-in feature, no existing codebase will break due to this change.

The code above shows that this is not true, because it broke due to this change. AFAICT the type narrowing behavior in old versions of pyright is safe, because old version of `pyright` don't allow you to assign a dict `d` to some TypedDict `D` if `d` has keys in it that aren't in `D`. E.g. the following code is flagged by 1.1.400

```py
class D(TypedDict):
    a: str

d: D = {"a": "b", "c": "d"}
```

**Comment by erictraut** [comment]
Pyright's new behavior is correct. Version 1.1.401 addressed an existing bug in the type narrowing code. This is independent of whether PEP 728 is eventually accepted.

If we extend your first code sample, we can see why it's unsound.

```python
class C(TypedDict):
    a: str
    b: str
    bb: str

c = C(a="a", b="b", bb="bb")
fn(c) # Crash
```

When TypedDict was introduced to the type system in [PEP 589](https://peps.python.org/pep-0589/), the authors were very clear in the spec that a value that is assignable to a TypedDict may contain additional keys that are not specified in the TypedDict definition. A type checker cannot make assumptions about the presence of additional keys or the types of their associated values. In other words, a TypedDict has historically been "open".

If PEP 728 is accepted, it will allow you to specify that a TypedDict is closed. For backward compatibility reasons, if a TypedDict is not explicitly declared as closed, it is assumed to be open.

**Comment by kylebebak** [comment]
Thank you for explaining that, I finally see the bug that was addressed here

`c` is assignable to `A | B`, because it's assignable to `B`. It's not assignable to `A`, so narrowing its type to `A` based on the presence of the `a` key is unsound

If PEP 728 is accepted (or we enable experimental behavior in 1.1.401) we can define A and B as closed, and `c` is not assignable to `A | B`, which prevent us from passing c into `fn` and hitting that crash
